clients/src/main/java/org/apache/kafka/common/requests/ProduceResponse.java

```diff
@@ -204,118 +75,79 @@ public ProduceResponse(Map<TopicPartition, PartitionResponse> responses) {
      * @param throttleTimeMs Time in milliseconds the response was throttled
      */
     public ProduceResponse(Map<TopicPartition, PartitionResponse> responses, int throttleTimeMs) {
-        this.responses = responses;
-        this.throttleTimeMs = throttleTimeMs;
+        this(new ProduceResponseData()
+            .setResponses(responses.entrySet()
```

Comment:
As we care of performances here, I wonder if we should try not using the stream api here.

Another trick would be to  turn `TopicProduceResponse` in the `ProduceResponse` schema into a map by setting `"mapKey": true` for the topic name. This would allow to iterate over `responses`, get or create `TopicProduceResponse` for the topic, and add the `PartitionProduceResponse` into it.

It may be worth trying different implementation to compare their performances.

Reply:
> It may be worth trying different implementation to compare their performances.

As we all care for performance, I'm ok to say goodbye to stream api :)

Reply:
I have addressed your suggestion and it does improve the performance.